### PR TITLE
feat(firebase_ui_auth): Add "Go Back" button to EmailLinkSignInView

### LINKED ISSUE #81: Method to back out from EmailLinkSignInScreen in iOS
### Discussed in https://github.com/firebase/FirebaseUI-Flutter/discussions/80

<div type='discussions-op-text'>

<sup>Originally posted by **rstieger** July 25, 2023</sup>
Given an app using SignInScreen from firebase_ui_auth, with EmailLinkAuthProvider as one of the providers:

A user might accidentally click on  the "Sign in with magic link" button. Or they might mistype their email address. In either case, on iOS there is no way to go back to the original SignInScreen, short of terminating and restarting the app. (Android always has a back button available.)

It would be better if EmailLinkSignInScreen/EmailLinkSignInView had the same "Go back" option that other screens have (for example, SMSCodeInputScreen). This could be a boolean parameter to the EmailLinkSignInScreen constructor in case some developers prefer the flow as-is.</div>

### PR DESCRIPTION
**Pull request by rstieger**
## Description

This PR adds a "Go back" button to EmailLinkSignInView. Other views and screens in firebase_ui_auth have
a similar button already, but EmailLinkSignInView does not, so users can get stuck if they got there by mistake,
mistyped their email address, etc. This button code was copied from SMSCodeInputScreen, so it should be
consistent with other auth flows in the package.

## Related Issues

Fixes #81

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for _all_ changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] All unit tests pass (`melos run test:unit:all` doesn't fail).
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is _not_ a breaking change.

[issue database]: https://github.com/firebase/FirebaseUI-Flutter/issues
[Contributor Guide]: https://github.com/firebase/FirebaseUI-Flutter/blob/main/docs/contributing.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_ui_auth/lib/src/views/email_link_sign_in_view.dart [hunk 1]
```diff
@@ -83,6 +83,14 @@ class _EmailLinkSignInViewState extends State<EmailLinkSignInView> {
               ),
             ],
             const SizedBox(height: 8),
+            UniversalButton(
```
Comment (lesnitsky) [inline packages/firebase_ui_auth/lib/src/views/email_link_sign_in_view.dart:86]: If this is used as a root view, `pop()` will close the app. We could have this configurable via widget property (e.g. `showBackButton` or use `Navigator.canPop`

**Review by lesnitsky — CHANGES_REQUESTED** [review]

File: packages/firebase_ui_auth/lib/src/views/email_link_sign_in_view.dart (hunk 1 above)
Comment (lesnitsky) [reply packages/firebase_ui_auth/lib/src/views/email_link_sign_in_view.dart:86]: hm, I think most of the views that have this button without any conditions are indeed not supposed to be used as a root view, but if you think some could – feel free to make a similar change.

Comment (rstieger) [reply packages/firebase_ui_auth/lib/src/views/email_link_sign_in_view.dart]: Only displaying this button if Navigator.canPop is true. Confirmed with my local app that if I push() a route to the EmailLinkSignInScreen then the "Go Back" button is shown, but if I pushReplacement() the same route (making it the root view) then "Go Back" is not shown.

File: packages/firebase_ui_auth/lib/src/views/email_link_sign_in_view.dart [hunk 2]
```diff
@@ -82,6 +82,16 @@ class _EmailLinkSignInViewState extends State<EmailLinkSignInView> {
                 },
               ),
             ],
+            if (Navigator.canPop(context)) ...[
```
Comment (lesnitsky) [inline packages/firebase_ui_auth/lib/src/views/email_link_sign_in_view.dart:85]: I think a single check during state init is enough, no need to check on every rebuild. `late final canPop = Navigator.canPop(context);` defined on the state should do.

**Review by lesnitsky — CHANGES_REQUESTED** [review]
Please also add a unit widget test verifying both scenarios.

File: packages/firebase_ui_auth/test/views/email_link_sign_in_view_test.dart [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  late EmailLinkAuthProvider emailLinkProvider;
+
+  setUp(() {
+    auth = MockAuth();
+    dynamicLinks = MockDynamicLinks();
+    final actionCodeSettings = ActionCodeSettings(
+      url: 'https://example.com',
+    );
+    emailLinkProvider = EmailLinkAuthProvider(
+        actionCodeSettings: actionCodeSettings, dynamicLinks: dynamicLinks);
```
Comment (lesnitsky) [inline packages/firebase_ui_auth/test/views/email_link_sign_in_view_test.dart:23]: trailing comma

File: packages/firebase_ui_auth/test/views/email_link_sign_in_view_test.dart [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            builder: (context) => TextButton(
+                child: const Text("Push"),
+                onPressed: () => Navigator.push(
+                    context,
+                    MaterialPageRoute(
+                      builder: (context) => Scaffold(
+                        body: EmailLinkSignInView(
+                            provider: emailLinkProvider, auth: auth),
+                      ),
+                    ))))));
```
Comment (lesnitsky) [inline packages/firebase_ui_auth/test/views/email_link_sign_in_view_test.dart:51]: Add trailing commas for nicer formatting

**Review by lesnitsky — CHANGES_REQUESTED** [review]

File: packages/firebase_ui_auth/test/views/email_link_sign_in_view_test.dart
Comment (lesnitsky) [inline packages/firebase_ui_auth/test/views/email_link_sign_in_view_test.dart:1]: Missing license header. Please run `melos run add-license-header`.

**Review by lesnitsky — CHANGES_REQUESTED** [review]

File: packages/firebase_ui_auth/test/views/email_link_sign_in_view_test.dart (hunk 4 above)
Comment (lesnitsky) [reply packages/firebase_ui_auth/test/views/email_link_sign_in_view_test.dart:59]:
```suggestion
                    ),
                  ),
                ),
              ),
            ),
          );
```

**Review by lesnitsky — APPROVED** [review]

**Pull request merged**